lib/membrane_rtmp_plugin/rtmp/sink/sink.ex

```diff
@@ -114,70 +128,63 @@ defmodule Membrane.RTMP.Sink do
   def handle_write(
         pad,
         buffer,
-        ctx,
+        _ctx,
         %{ready: true, buffered_frame: {frame_pad, frame}} = state
       ) do
-    {_, state} = handle_write(frame_pad, frame, ctx, Map.delete(state, :buffered_frame))
-    {_, state} = handle_write(pad, buffer, ctx, state)
-    {{:ok, demand: state.current_timestamps |> Enum.min_by(&elem(&1, 1)) |> elem(0)}, state}
+    state = write_frame(state, frame_pad, frame) |> write_frame(pad, buffer)
+    {{:ok, demand: get_demand(state)}, Map.put(state, :buffered_frame, nil)}
   end
 
   @impl true
-  def handle_write(:video, buffer, _ctx, %{ready: true} = state) do
-    case Native.write_video_frame(
-           state.native,
-           buffer.payload,
-           buffer.dts,
-           buffer.metadata.h264.key_frame?
-         ) do
-      {:ok, native} ->
-        state = Map.put(state, :native, native)
-
-        state =
-          Map.update(state, :current_timestamps, %{video: buffer.dts}, fn curr_tmps ->
-            Map.put(curr_tmps, :video, buffer.dts)
-          end)
-
-        {{:ok, demand: state.current_timestamps |> Enum.min_by(&elem(&1, 1)) |> elem(0)}, state}
+  def handle_write(pad, buffer, _ctx, state) do
+    state = write_frame(state, pad, buffer)
+    {{:ok, demand: get_demand(state)}, state}
+  end
 
-      {:error, reason} ->
-        raise("Writing video frame failed with reason: #{reason}")
+  @impl true
+  def handle_end_of_stream(pad, ctx, state) do
+    if ctx.pads |> Map.values() |> Enum.map(& &1.end_of_stream?) |> Enum.all?() do
```

Comment:
Does it work? I thought that context will be updated with eos for the current pad after this callback is called. Also, you overcomplicated a little bit
```suggestion
    if ctx.pads |> Map.delete(pad) |> Map.values() |> Enum.all?(& &1.end_of_stream?) do
```

Reply:
After running experiments with detailed logging it appears that int the case of handle_end_of_stream context is updated prior to callback execution. Nevertheless, I will remove unnecessary Enum.map as you suggested.